File: sdk/c/src/providers/chatgpt_org_uk.c
/**
 * mail.chatgpt.org.uk 渠道实现
 */
#include "tempmail_internal.h"

#define CG_BASE "https://mail.chatgpt.org.uk/api"
#define CG_HOME "https://mail.chatgpt.org.uk/"

static const char* cg_headers[] = {
    "Accept: */*",
    "Referer: https://mail.chatgpt.org.uk/",
    "Origin: https://mail.chatgpt.org.uk",
    "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36",
    "DNT: 1",
    NULL
};

static char* cg_extract_gm_sid(const char *cookies) {
    if (!cookies) return NULL;
    const char *start = strstr(cookies, "gm_sid=");
    if (!start) return NULL;
    const char *end = strchr(start, ';');
    size_t len = end ? (size_t)(end - start) : strlen(start);
    char *out = (char*)malloc(len + 1);
    if (!out) return NULL;
    memcpy(out, start, len);
    out[len] = '\0';
    return out;
}

static char* cg_fetch_gm_sid_once(long *status_out) {
    if (status_out) *status_out = 0;
    tm_http_response_t *home = tm_http_request(TM_HTTP_GET, CG_HOME, cg_headers, NULL, 15);
    if (!home) return NULL;
    if (status_out) *status_out = home->status;
    if (home->status != 200) { tm_http_response_free(home); return NULL; }

    char *gm_sid = cg_extract_gm_sid(home->cookies);
    tm_http_response_free(home);
    return gm_sid;
}

static char* cg_fetch_gm_sid(void) {
    long status = 0;
    char *gm_sid = cg_fetch_gm_sid_once(&status);
    if (gm_sid) return gm_sid;
    if (status == 401 || status == 200) {
        return cg_fetch_gm_sid_once(&status);
    }
    return NULL;
}

static char* cg_fetch_inbox_token_once(const char *email, const char *gm_sid, long *status_out) {
    if (status_out) *status_out = 0;
    char body[512];
    snprintf(body, sizeof(body), "{\"email\":\"%s\"}", email);

    char cookie_hdr[512];
    snprintf(cookie_hdr, sizeof(cookie_hdr), "Cookie: %s", gm_sid);

    const char* headers[] = {
        "Accept: */*",
        "Referer: https://mail.chatgpt.org.uk/",
        "Origin: https://mail.chatgpt.org.uk",
        "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36",
        "DNT: 1",
        "Content-Type: application/json",
        cookie_hdr,
        NULL
    };

    tm_http_response_t *resp = tm_http_request(TM_HTTP_POST, CG_BASE "/inbox-token", headers, body, 15);
    if (!resp) return NULL;
    if (status_out) *status_out = resp->status;
    if (resp->status != 200) { tm_http_response_free(resp); return NULL; }

    cJSON *json = cJSON_Parse(resp->body);
    tm_http_response_free(resp);
    if (!json) return NULL;

    cJSON *auth = cJSON_GetObjectItemCaseSensitive(json, "auth");
    const char *token = cJSON_GetStringValue(cJSON_GetObjectItemCaseSensitive(auth, "token"));
    if (!token || !token[0]) { cJSON_Delete(json); return NULL; }

    char *out = tm_strdup(token);
    cJSON_Delete(json);
    return out;
}

static char* cg_fetch_inbox_token(const char *email) {
    long status = 0;
    char *gm_sid = cg_fetch_gm_sid();
    if (!gm_sid) return NULL;

    char *token = cg_fetch_inbox_token_once(email, gm_sid, &status);
    if (!token && status == 401) {
        free(gm_sid);
        gm_sid = cg_fetch_gm_sid();
        if (!gm_sid) return NULL;
        token = cg_fetch_inbox_token_once(email, gm_sid, &status);
    }
    free(gm_sid);
    return token;
}

static tm_email_t* cg_fetch_emails_once(const char *token, const char *email, int *count, long *status_out) {
    if (count) *count = -1;
    if (status_out) *status_out = 0;
    if (!token || !token[0]) return NULL;

    char url[512];
    snprintf(url, sizeof(url), CG_BASE "/emails?email=%s", email);

    char token_hdr[1024];
    snprintf(token_hdr, sizeof(token_hdr), "x-inbox-token: %s", token);

    const char* headers[] = {
        "Accept: */*",
        "Referer: https://mail.chatgpt.org.uk/",
        "Origin: https://mail.chatgpt.org.uk",
        "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36",
        "DNT: 1",
        token_hdr,
        NULL
    };

    tm_http_response_t *resp = tm_http_request(TM_HTTP_GET, url, headers, NULL, 15);
    if (!resp) return NULL;
    if (status_out) *status_out = resp->status;
    if (resp->status != 200) { tm_http_response_free(resp); return NULL; }

    cJSON *json = cJSON_Parse(resp->body);
    tm_http_response_free(resp);
    if (!json) return NULL;

    if (!cJSON_IsTrue(cJSON_GetObjectItemCaseSensitive(json, "success"))) { cJSON_Delete(json); return NULL; }

    cJSON *data = cJSON_GetObjectItemCaseSensitive(json, "data");
    cJSON *arr = cJSON_GetObjectItemCaseSensitive(data, "emails");
    int n = cJSON_IsArray(arr) ? cJSON_GetArraySize(arr) : 0;
    if (count) *count = n;
    if (n == 0) { cJSON_Delete(json); return NULL; }

    tm_email_t *emails = tm_emails_new(n);
    for (int i = 0; i < n; i++) emails[i] = tm_normalize_email(cJSON_GetArrayItem(arr, i), email);
    cJSON_Delete(json);
    return emails;
}


tm_email_info_t* tm_provider_chatgpt_org_uk_generate(void) {
    tm_http_response_t *resp = tm_http_request(TM_HTTP_GET, CG_BASE "/generate-email", cg_headers, NULL, 15);
    if (!resp || resp->status != 200) { tm_http_response_free(resp); return NULL; }

    cJSON *json = cJSON_Parse(resp->body);
    tm_http_response_free(resp);
    if (!json) return NULL;

    if (!cJSON_IsTrue(cJSON_GetObjectItemCaseSensitive(json, "success"))) { cJSON_Delete(json); return NULL; }

    cJSON *data = cJSON_GetObjectItemCaseSensitive(json, "data");
    const char *email = cJSON_GetStringValue(cJSON_GetObjectItemCaseSensitive(data, "email"));
    if (!email || !email[0]) { cJSON_Delete(json); return NULL; }

    char *token = cg_fetch_inbox_token(email);
    if (!token) { cJSON_Delete(json); return NULL; }

    tm_email_info_t *info = tm_email_info_new();
    info->channel = CHANNEL_CHATGPT_ORG_UK;
    info->email = tm_strdup(email);
    info->token = token;
    cJSON_Delete(json);
    return info;
}


tm_email_t* tm_provider_chatgpt_org_uk_get_emails(const char *token, const char *email, int *count) {
    *count = -1;
    if (!token || !token[0]) return NULL;

    long status = 0;
    tm_email_t *emails = cg_fetch_emails_once(token, email, count, &status);
    if (status == 401) {
        char *refreshed = cg_fetch_inbox_token(email);
        if (refreshed) {
            emails = cg_fetch_emails_once(refreshed, email, count, &status);
            free(refreshed);
        }
    }
    return emails;
}

File: sdk/rust/src/providers/chatgpt_org_uk.rs
/*!
 * mail.chatgpt.org.uk 渠道实现
 */

use serde_json::Value;
use crate::types::{Channel, EmailInfo, Email};
use crate::normalize::normalize_email;
use crate::config::{http_client, block_on, get_current_ua};

const BASE_URL: &str = "https://mail.chatgpt.org.uk/api";
const HOME_URL: &str = "https://mail.chatgpt.org.uk/";

fn is_401_error(message: &str) -> bool {
    message.contains("401")
}

fn is_gm_sid_error(message: &str) -> bool {
    message.to_lowercase().contains("gm_sid")
}

fn fetch_gm_sid_once() -> Result<String, String> {
    block_on(async {
        let resp = http_client()
            .get(HOME_URL)
            .header("User-Agent", get_current_ua())
            .header("Accept", "*/*")
            .header("Referer", HOME_URL)
            .header("Origin", "https://mail.chatgpt.org.uk")
            .header("DNT", "1")
            .send().await.map_err(|e| format!("chatgpt-org-uk home request failed: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("chatgpt-org-uk home failed: {}", resp.status()));
        }

        let gm_sid = resp.headers().get_all("set-cookie")
            .iter()
            .filter_map(|v| v.to_str().ok())
            .find_map(|s| {
                let part = s.split(';').next().unwrap_or("");
                if part.starts_with("gm_sid=") {
                    Some(part["gm_sid=".len()..].to_string())
                } else {
                    None
                }
            })
            .ok_or("Failed to extract gm_sid cookie")?;

        Ok(gm_sid)
    })
}

fn fetch_gm_sid() -> Result<String, String> {
    match fetch_gm_sid_once() {
        Ok(gm_sid) => Ok(gm_sid),
        Err(err) => {
            if is_401_error(&err) || is_gm_sid_error(&err) {
                return fetch_gm_sid_once();
            }
            Err(err)
        }
    }
}

fn fetch_inbox_token_once(email: &str, gm_sid: &str) -> Result<String, String> {
    let email = email.to_string();
    let gm_sid = gm_sid.to_string();
    block_on(async {
        let resp = http_client()
            .post(format!("{}/inbox-token", BASE_URL))
            .header("User-Agent", get_current_ua())
            .header("Accept", "*/*")
            .header("Referer", HOME_URL)
            .header("Origin", "https://mail.chatgpt.org.uk")
            .header("DNT", "1")
            .header("Content-Type", "application/json")
            .header("Cookie", format!("gm_sid={}", gm_sid))
            .json(&serde_json::json!({"email": email}))
            .send().await.map_err(|e| format!("chatgpt-org-uk inbox-token request failed: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("chatgpt-org-uk inbox-token failed: {}", resp.status()));
        }

        let data: Value = resp.json().await.map_err(|e| format!("parse failed: {}", e))?;
        let token = data["auth"]["token"].as_str().unwrap_or("");
        if token.is_empty() {
            return Err("Failed to get inbox token".into());
        }

        Ok(token.to_string())
    })
}

fn fetch_inbox_token(email: &str) -> Result<String, String> {
    let gm_sid = fetch_gm_sid()?;
    match fetch_inbox_token_once(email, &gm_sid) {
        Ok(token) => Ok(token),
        Err(err) => {
            if is_401_error(&err) {
                let gm_sid = fetch_gm_sid()?;
                return fetch_inbox_token_once(email, &gm_sid);
            }
            Err(err)
        }
    }
}

pub fn generate_email() -> Result<EmailInfo, String> {
    let email = block_on(async {
        let resp = http_client()
            .get(format!("{}/generate-email", BASE_URL))
            .header("User-Agent", get_current_ua())
            .header("Accept", "*/*")
            .header("Referer", HOME_URL)
            .header("Origin", "https://mail.chatgpt.org.uk")
            .header("DNT", "1")
            .send().await.map_err(|e| format!("chatgpt-org-uk request failed: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("chatgpt-org-uk generate failed: {}", resp.status()));
        }

        let data: Value = resp.json().await.map_err(|e| format!("parse failed: {}", e))?;
        if !data["success"].as_bool().unwrap_or(false) {
            return Err("Failed to generate email".into());
        }

        let email = data["data"]["email"].as_str().unwrap_or("").to_string();
        if email.is_empty() {
            return Err("Failed to generate email".into());
        }

        Ok(email)
    })?;

    let token = fetch_inbox_token(&email)?;

    Ok(EmailInfo {
        channel: Channel::ChatgptOrgUk,
        email,
        token: Some(token),
        expires_at: None,
        created_at: None,
    })
}

fn fetch_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let token = token.to_string();
    let email = email.to_string();
    block_on(async {
        let resp = http_client()
            .get(format!("{}/emails?email={}", BASE_URL, urlencoding::encode(&email)))
            .header("User-Agent", get_current_ua())
            .header("Accept", "*/*")
            .header("Referer", HOME_URL)
            .header("Origin", "https://mail.chatgpt.org.uk")
            .header("DNT", "1")
            .header("x-inbox-token", token)
            .send().await.map_err(|e| format!("chatgpt-org-uk request failed: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("chatgpt-org-uk get emails failed: {}", resp.status()));
        }

        let data: Value = resp.json().await.map_err(|e| format!("parse failed: {}", e))?;
        if !data["success"].as_bool().unwrap_or(false) {
            return Err("Failed to get emails".into());
        }

        Ok(data["data"]["emails"].as_array()
            .map(|arr| arr.iter().map(|raw| normalize_email(raw, &email)).collect())
            .unwrap_or_default())
    })
}

pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    if token.is_empty() {
        return Err("token is required for chatgpt-org-uk".into());
    }

    match fetch_emails(token, email) {
        Ok(emails) => Ok(emails),
        Err(err) => {
            if is_401_error(&err) {
                let refreshed = fetch_inbox_token(email)?;
                return fetch_emails(&refreshed, email);
            }
            Err(err)
        }
    }
}
